guides/keras_nlp/neural_machine_translation.py

```diff
@@ -0,0 +1,498 @@
+when the input sequence length is shorter than the maximum sequence length.
+- [UNK] - Unknown token.
+- [START] - Token that marks the start of the input sequence.
+- [END] - Token that marks the end of the input sequence.
+"""
+
+reserved_tokens = ["[PAD]", "[UNK]", "[START]", "[END]"]
+
+eng_samples = [text_pair[0] for text_pair in train_pairs]
+eng_vocab = train_word_piece(eng_samples, ENG_VOCAB_SIZE, reserved_tokens)
+
+spa_samples = [text_pair[1] for text_pair in train_pairs]
+spa_vocab = train_word_piece(spa_samples, SPA_VOCAB_SIZE, reserved_tokens)
+
+"""
+Let's see some tokens!
+"""
+print("English Tokens: ", eng_vocab[100:110])
+print("Spanish Tokens: ", spa_vocab[100:110])
+
+"""
+Now, let's define the tokenizers. We will use the vocabularies obtained above as
+input to the tokenizers. We will define a maximum sequence length so that
+all sequences are padded to the same length, if the length of the sequence is
+less than the specified sequence length. Otherwise, the sequence is truncated.
+"""
+
+eng_tokenizer = keras_nlp.tokenizers.WordPieceTokenizer(
+    vocabulary=eng_vocab,
+    lowercase=False,
+    split_pattern=" ",
+    sequence_length=MAX_SEQUENCE_LENGTH,
+)
+spa_tokenizer = keras_nlp.tokenizers.WordPieceTokenizer(
+    vocabulary=spa_vocab,
+    lowercase=False,
+    split_pattern=" ",
+    sequence_length=MAX_SEQUENCE_LENGTH + 1,
+)
+
+"""
+Let's try and tokenize a sample from our dataset! To verify whether the text has
+been tokenized correctly, we can also detokenize the list of tokens back to the
+original text.
+"""
+
+eng_input_ex = text_pairs[0][0]
+eng_tokens_ex = eng_tokenizer.tokenize(eng_input_ex)
+print("English sentence: ", eng_input_ex)
+print("Tokens: ", eng_tokens_ex)
+print("Recovered text after detokenizing: ", eng_tokenizer.detokenize(eng_tokens_ex))
+
+print()
+
+spa_input_ex = text_pairs[0][1]
+spa_tokens_ex = spa_tokenizer.tokenize(spa_input_ex)
+print("Spanish sentence: ", spa_input_ex)
+print("Tokens: ", spa_tokens_ex)
+print("Recovered text after detokenizing: ", spa_tokenizer.detokenize(spa_tokens_ex))
+
+"""
+## Format Datasets
+
+Next, we'll format our datasets.
+
+At each training step, the model will seek to predict target words N+1 (and beyond)
+using the source sentence and the target words 0 to N.
+
+As such, the training dataset will yield a tuple `(inputs, targets)`, where:
+
+- `inputs` is a dictionary with the keys `encoder_inputs` and `decoder_inputs`.
+`encoder_inputs` is the tokenized source sentence and `decoder_inputs` is the target sentence "so far",
+that is to say, the words 0 to N used to predict word N+1 (and beyond) in the target sentence.
+- `target` is the target sentence offset by one step:
+it provides the next words in the target sentence -- what the model will try to predict.
+
+Before we tokenize the text, we will add [START] and [END] tokens to the input
+Spanish sentence.
+"""
+
+train_pairs = [
```

Comment:
One bigger comment for this guide and the other guide. Instead of doing string manipulation like this, which is fragile and cumbersome, let's add the start and end tokens after tokenization.

If you do the `format_dataset` call before batching, this should be pretty simple. Just concat `eng_tokenize.token_to_id("[START]")`, `eng_tokenizer(eng)`, and `eng_tokenize.token_to_id("[END]")`

Reply:
@mattdangerw, two questions:
1. Don't we also want to show the tokenizer being called on a batched dataset? Won't users mostly use the tokenizer on a batch rather than a sample?
2. We can't concatenate `eng_tokenize.token_to_id("[START]")`, `eng_tokenizer(eng)`, and `eng_tokenize.token_to_id("[END]")` directly. Rather, we'd have to insert the `[END]` token right before the first occurrence of `[PAD]` (if padding tokens are present). If padding tokens are not present, we can just add `[END]` at the end. Right?